oeps/oep-0032-arch-unique-identifier-for-users.rst

```diff
@@ -0,0 +1,153 @@
+
+
+Context
+=======
+
+In 2016, the analytics team wrote a decision to use the the LMS user_id in `Using consistent user identifiers in Segment events`_.
+
+In addition to needing a consistent user identifier for internal events, we also need one for public events, REST APIs and JavaScript APIs.
+
+.. _Using consistent user identifiers in Segment events: https://openedx.atlassian.net/wiki/spaces/AN/pages/144441849/Using+consistent+user+identifiers+in+Segment+events
+
+
+User Identifier Requirements
+----------------------------
+
+The requirements for a User Identifier as described by the analytics decision document, are also applicable for our REST APIs and JavaScript APIs.
+
+* Uniqueness: the identifier must not represent more than one user.
+
+* The identifier must map back to other information about the user (e.g. demographics, enrollment, etc.).
+
+* The identifier for a given user should not change.
+
+* The identifier should not contain `Personally Identifiable Information (PII)`_.
+
+* The identifier should be consistent across all systems, including the LMS and Studio, mobile apps for iOS and Android, and other Open edX Services like Ecommerce.
+
+.. _Personally Identifiable Information (PII): oep-0030-arch-pii-markup-and-auditing.rst
+
+
+User Identifier Candidate Comparison
+------------------------------------
+
+The following table summarizes the comparisons of a list of candidate user identifiers.
+
+.. list-table::
+   :header-rows: 1
+   :widths: 25 75
+
+   * - Identifier
+     - Limitations
+   * - username
+     - Username is considered PII.
+   * - LMS user_id
+     - Information leakage of auth_user table with easily guessable values; tied down to implementation of the housing database.
+   * - email address
+     - Is PII and modifiable by the user.
+   * - anonymous user_id
+     - It is currently constructed by hashing the user's LMS user_id with the Django server's *SECRET_KEY* value. This value will change when the *SECRET_KEY* is rotated.
+
+       *Note*: Although this id was named the "anonymous user_id", it is no more anonymous than the "LMS user_id" from a PII perspective.
+
+As noted above, the LMS user_id was selected for analytics as the best match given the requirements.
+
+Drawbacks of the LMS user_id
+----------------------------
+
+As noted in the `User Identifier Candidate Comparison`_ above, using the LMS user_id has the following two drawbacks:
+
+* Information leakage of auth_user table with easily guessable values.
+
+* The id is tied down to implementation of the housing database.
+
+Cost of an alternative to the LMS user_id
+-----------------------------------------
+
+If we were to try to eliminate the drawbacks of using the auto-incremented LMS user_id as our unique identifier, an alternative might be to introduce a new id, like a UUID (Universally Unique IDentifier), that did not have the `Drawbacks of the LMS user_id`_.
```

Comment:
What about sticking with the existing `user_id` column, but changing the insertion algorithm to generate a random user ID instead of a sequential one for new users? Obviously that wouldn't help existing users, but that would significantly reduce this drawback for new ones as well as new Open edX installations.

Reply:
Hi @bradenmacdonald. I had thought of this, but I was under the understanding that the column and all foreign-keys across the entire system would need to change size to binary(16). That doesn't seem small.  And it ignores the fact that you'd have additional questions about when and where to use the human readable form of the UUID, and other complications of dealing with proper indexing, etc.  Thoughts?

Reply:
The column is currently a 32-bit integer. I agree that changing it to a 64-bit int or a 128-bit UUID would be prohibitively difficult (due to the foreign keys and the size of auth_user; you'd need to use pt-online-schema-change for the large tables and it would take ages).

However, what I'm suggesting is just keep the schema unchanged (ID remains a 32-bit integer) but when there's a new user, assign them a random 32-bit integer as their ID, rather than the next integer in the auto_increment sequence. (Plus some logic to deal with conflicts so insertions never fail).

Reply:
I see.  Nothing in this OEP prohibits this idea if someone wants to take that on.  The major disadvantage for UUIDs is a potential performance issue with the index because the values are not sorted.  There are solutions like [rearranging the time-related bits](https://mysqlserverteam.com/mysql-8-0-uuid-support/) in the MYSQL solution.  Someone would have to find the right way to make something like this work with 32-bits for it to ever work for edx.org.

I could acknowledge this idea, but I don't want to over explain it and I don't want to seem like anything is getting promised in this OEP along that front.

Reply:
@bradenmacdonald: I went for simple on the update, so I did not mention this.  As noted, I don't think the decision states anything against this possibility.